src/services/rgbpp.ts

```diff
@@ -298,9 +306,87 @@ export default class RgbppCollector extends BaseQueueWorker<IRgbppCollectRequest
   }
 
+  async isIsomorphicTx(
+    btcTx: Transaction,
+    ckbTx: CKBComponents.RawTransaction,
+    validateCommitment?: boolean,
+  ): Promise<boolean> {
+    // Find the commitment from the btc_tx
+    const btcTxCommitment = tryGetCommitmentFromBtcTx(btcTx);
+    if (!btcTxCommitment) {
+      return false;
+    }
+
+    // Check inputs:
+    // 1. Find the last index of the type inputs
+    // 2. Check if all rgbpp_lock inputs can be found in the btc_tx.vin
+    // 3. Check if the inputs contain at least one rgbpp_lock cell (as L1-L1 and L1-L2 transactions should have)
+    const inputs = await this.cradle.ckb.getInputCellsByOutPoint(ckbTx.inputs.map((input) => input.previousOutput!));
+    const lastTypeInputIndex = findLastIndex(inputs, (input) => !!input.cellOutput.type);
+    const anyRgbppLockInput = inputs.some((input) => isRgbppLock(input.cellOutput.lock));
+    if (!anyRgbppLockInput) {
+      return false;
+    }
+    const allInputsValid = inputs.every((input, index) => {
+      if (!input.cellOutput.type) {
+        return true;
+      }
+      if (!isRgbppLock(input.cellOutput.lock)) {
+        return true;
+      }
+      const btcInput = btcTx.vin[index];
+      const rgbppLockArgs = unpackRgbppLockArgs(input.cellOutput.lock.args);
+      return btcInput && btcInput.txid === remove0x(rgbppLockArgs.btcTxid) && btcInput.vout === rgbppLockArgs.outIndex;
+    });
+    if (!allInputsValid) {
+      return false;
+    }
+
+    // Check outputs:
+    // 1. Find the last index of the type outputs, and check if at least one type output exists
+    // 2. Check if all type outputs are rgbpp_lock or btc_time_lock cells
+    // 4. Check if each rgbpp_lock cell has an isomorphic UTXO in the btc_tx.vout
+    // 5. Check if each btc_time_lock cell contains the corresponding btc_txid in the lock args
+    const lastTypeOutputIndex = findLastIndex(ckbTx.outputs, (output) => !!output.type);
+    if (lastTypeOutputIndex < 0) {
+      return false;
+    }
+    const anyRelatedLockToTypeOutput = ckbTx.outputs.some(
+      (output) => output.type && (isRgbppLock(output.lock) || isBtcTimeLock(output.lock)),
+    );
+    if (!anyRelatedLockToTypeOutput) {
+      return false;
+    }
+    const allOutputsValid = ckbTx.outputs.every((output) => {
+      if (isRgbppLock(output.lock)) {
+        const rgbppLockArgs = unpackRgbppLockArgs(output.lock.args);
+        const btcTxId = remove0x(rgbppLockArgs.btcTxid);
+        if (btcTxId !== RGBPP_TX_ID_PLACEHOLDER && (btcTxId !== btcTx.txid || !btcTx.vout[rgbppLockArgs.outIndex])) {
+          return false;
+        }
+      }
+      if (isBtcTimeLock(output.lock)) {
+        const btcTxId = remove0x(btcTxIdFromBtcTimeLockArgs(output.lock.args));
+        if (btcTxId !== RGBPP_TX_ID_PLACEHOLDER && btcTx.txid !== btcTxId) {
+          return false;
+        }
+      }
+      return true;
+    });
+    if (!allOutputsValid) {
+      return false;
+    }
+
+    // Compare commitment between btc_tx and ckb_tx
```

Comment:
The overall goal of the transaction is to compare one BTC_TX with one CKB_TX, I assumed it may or may not be an on-chain transaction, so the function receives a `RawTransaction`. Do you think we need to change the implementation here?

Reply:
> I assumed the CKB_TX may or may not be an on-chain transaction, so the function receives a `RawTransaction` instead of a `Transaction` directly.

I see.
will review the code logic again.

Reply:
https://github.com/ckb-cell/btc-assets-api/blob/e0c1c517b3079a5cec7d4633111f8406ce87c92f/src/services/rgbpp.ts#L348-L378

LGTM